Include tags from Kamon.environment.tags

CasperKoning opened the pull request:
Since https://github.com/kamon-io/Kamon/pull/515 it is now possible to provide custom tags in the `application.conf`:
```
kamon.environment.tags {
  custom1 = "test1"
  env = staging
}
```

With this PR, this reporter includes these tags as well.

CasperKoning committed:
Bump kamon-core version to latest

CasperKoning committed:
Include tags from Kamon.environment.tags

CasperKoning commented:
Note that with this change, the extraction of the tags from the environment is always on. I could include a conditional check to maybe not include them here https://github.com/kamon-io/kamon-prometheus/pull/10/files#diff-af95be89e23c393025b86a7ed8ae991cR60

This could be turned on or off via the `application.conf`. What would then be the preferred default (on/off)?

japgolly commented on src/main/scala/kamon/prometheus/ScrapeDataBuilder.scala:
```diff
@@ -140,9 +141,10 @@ class ScrapeDataBuilder(prometheusConfig: PrometheusReporter.Configuration) {
   private def appendTags(tags: Map[String, String]): Unit = {
-    if(tags.nonEmpty) append("{")
+    val allTags = environmentTags ++ tags
+    if(allTags.nonEmpty) append("{")
```
Better performance would be:
```scala
val tagIterator = environmentTags.iterator ++ tags.iterator
val addBraces = tagIterator.hasNext()
if (addBraces) append("{")
// ...
```

japgolly replied:
Iterator `++` much more efficient than Map `++` followed by `.iterator`

CasperKoning replied:
The results are potentially different though, as concatenating iterators may yield duplicate tags, while concatenating maps ensures that there's only one entry per key.

Exposing multiple labels to Prometheus won't break the Prometheus scraper, but will lead to Prometheus only keeping the one that occurs later in the sequence of labels. 

Personally, I'd opt for having an explicit preference in this exporter instead of relying on Prometheus to do the deduplication. That does bring up the following point: which tag would we prefer when we run into a duplicate? The one from the `Kamon.environment`, or the one from the regular `tags`?

ivantopo replied:
In the case above I would say that environment tag should win. Typically this would be used to discern the instance from which data is being collected so all tags for a metric should have the very same tags, it would me a mistake to provide a custom tag that overrides the environment tags.

Thanks for bringing this up, I'll make this makes it to the docs! (see https://github.com/kamon-io/kamon.io/issues/43)

CasperKoning committed:
Make including environment tags as labels configurable

To keep this exporter working in a backwards consistent way, We
turn the inclusion off by default. That is, whenever somebody
upgrades to a newer version of this project, the metrics don't
suddenly include more labels.

CasperKoning commented:
I included a new commit to exclude the `Kamon.environment.tags` by default to stay consistent with previous behavior. You can configure this via `application.conf`, i.e.
```
kamon.prometheus.include-environment-tags = no
```
(seems reasonable to add some doc in the README for this, which I'll probably do later)

This does show that things are getting a bit messy because `PrometheusReporter.Configuration` is both passed in to `ScrapeDataBuilder` as well as that it's being used outside to conditionally pass in the tags or not (we make that decision externally because we want to be able to test the `ScrapeDataBuilder` without having a dependency on the global `Kamon.environment.tags`).

Since the `PrometheusReporter.Configuration` contains some configs that are unrelated to `ScrapeDataBuilder`, it probably makes sense to refactor that a bit. For example, we could pass in a `ScrapeDataBuilder.Configuration` class that only included the relevant fields:
```
case class Configuration(
  defaultBuckets: Seq[java.lang.Double], 
  timeBuckets: Seq[java.lang.Double], 
  informationBuckets: Seq[java.lang.Double],
  environmentTags: Map[String, String]
)
```

ivantopo commented on src/test/scala/kamon/prometheus/ScrapeDataBuilderSpec.scala:
```diff
@@ -183,10 +183,27 @@ class ScrapeDataBuilderSpec extends WordSpec with Matchers {
+          |gauge_one_seconds{env_key="env_value"} 20.0
+        """.stripMargin.trim()
+      }
+    }
```
There should be one additional test covering the environment tags vs metric tags mentioned above.

ivantopo reviewed (changes requested):
@CasperKoning thanks a lot for the PR! Other than the missing test I think this is good to be merged.. if you have the time to add the test please do, otherwise let me know and I'll merge it and add the test.

I have some extra free time this weekend and would like to move as many things forward as possible :-)

CasperKoning committed:
Let environment tags have precedence over custom tags

CasperKoning committed:
Document feature for includig environment tags

CasperKoning commented:
I had some spare time so I decided to immediately jump on this 👍 

Let me know if there's anything that may need some more work. Maybe squash the commits into one?

ivantopo commented:
No need to squash, now Github offers it as an option! :tada:

Thanks a lot for this @CasperKoning!
